File: src/javascript/algorithms/dynamicProgramming/interleavingStrings/interleavingStrings.js
function isInterleave(s1, s2, s3) {
    const m = s1.length, n = s2.length;
    if (m + n !== s3.length) return false;
    const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(false));
    dp[m][n] = true;
    for (let i = m; i >= 0; i--) {
        for (let j = n; j >= 0; j--) {
            if (i === m && j === n) continue;
            const k = i + j;
            const fromS1 = i < m && s1[i] === s3[k] && dp[i + 1][j];
            const fromS2 = j < n && s2[j] === s3[k] && dp[i][j + 1];
            dp[i][j] = fromS1 || fromS2;
        }
    }
    return dp[0][0];
}

console.log("Input: s1='aab', s2='axy', s3='aaxaby' -> Output:", isInterleave('aab', 'axy', 'aaxaby')); // true
console.log("Input: s1='aab', s2='axy', s3='abaaxy' -> Output:", isInterleave('aab', 'axy', 'abaaxy')); // false
console.log("Input: s1='', s2='', s3='' -> Output:", isInterleave('', '', ''));
console.log("Input: s1='abc', s2='', s3='abc' -> Output:", isInterleave('abc', '', 'abc'));
console.log("Input: s1='', s2='xyz', s3='xyz' -> Output:", isInterleave('', 'xyz', 'xyz'));
console.log("Input: s1='abc', s2='def', s3='abcd' -> Output:", isInterleave('abc', 'def', 'abcd'));
console.log("Input: s1='aa', s2='ab', s3='aaba' -> Output:", isInterleave('aa', 'ab', 'aaba'));

File: src/python/algorithms/dynamicProgramming/interleavingStrings/interleaving_strings.py
def is_interleave(s1, s2, s3):
    m, n = len(s1), len(s2)
    if m + n != len(s3):
        return False
    dp = [[False] * (n + 1) for _ in range(m + 1)]
    dp[m][n] = True
    for i in range(m, -1, -1):
        for j in range(n, -1, -1):
            if i == m and j == n:
                continue
            k = i + j
            from_s1 = i < m and s1[i] == s3[k] and dp[i + 1][j]
            from_s2 = j < n and s2[j] == s3[k] and dp[i][j + 1]
            dp[i][j] = from_s1 or from_s2
    return dp[0][0]

if __name__ == "__main__":
    print("Input: s1='aab', s2='axy', s3='aaxaby' -> Output:", is_interleave('aab', 'axy', 'aaxaby')) # True
    print("Input: s1='aab', s2='axy', s3='abaaxy' -> Output:", is_interleave('aab', 'axy', 'abaaxy')) # False
    print("Input: s1='', s2='', s3='' -> Output:", is_interleave('', '', '')) # True
    print("Input: s1='abc', s2='', s3='abc' -> Output:", is_interleave('abc', '', 'abc')) # True
    print("Input: s1='', s2='xyz', s3='xyz' -> Output:", is_interleave('', 'xyz', 'xyz')) # True
    print("Input: s1='abc', s2='def', s3='abcd' -> Output:", is_interleave('abc', 'def', 'abcd')) # False
    print("Input: s1='aa', s2='ab', s3='aaba' -> Output:", is_interleave('aa', 'ab', 'aaba')) # True
